### PR TITLE
wip: Rwu/enh/runner

### PR DESCRIPTION
**Pull request by ipa-rwu**

### CODE REVIEW

File: araig_test_runners/scripts/test_4_static_obstacle [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+            else:
+                rospy.loginfo(rospy.get_name() + ": Robot is still moving, waiting for robot reaching goal and stop")
+                self.current_state = State.start_test
+        
+        # Test complete if robot doesn't move any more and reached the goal
+        if self.current_state == State.reached_goal:
+            rospy.loginfo(rospy.get_name() + ": Robot has reached goal, ending test in success!")
+            self._publishers["test_succeeded"].publish(self.buildNewBoolStamped(True))
+            self._publishers["test_completed"].publish(self.buildNewBoolStamped(True))
+            self.current_state = State.complete_test
```
Comment (ipa-rwu) [inline araig_test_runners/scripts/test_4_static_obstacle:107]: #20 
runner will keep checking if it gets *signal.calc.robot_reached_goal_and_stopped*

File: araig_test_runners/scripts/test_4_static_obstacle [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    def __init__(self, rate):
+        
+        extend_subscribers_dict = {
+            "goal"                  : "/signal/calc/robot_reached_goal_and_stopped",
+            "robot_in_collision"    : "/signal/calc/robot_in_collision"
+        }
+        extend_publishers_dict = {
+        }
+        extend_param_list = {
+            "timeout_check_goal",
```
Comment (ipa-kut) [inline araig_test_runners/scripts/test_4_static_obstacle:45]: can be renamed to `timeout_check_goal_oscillations`

File: araig_test_runners/scripts/test_4_static_obstacle [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        # If robot has stopped (because it found an obstacle), but maybe robot will still move
+        if self.current_state == State.start_test and self.getSafeFlag("goal"):
+            rospy.loginfo(rospy.get_name() + ": Robot has reached goal and stop, will check if robot move again in {}s".format(self.config_param['timeout_check_goal']))
+            self.current_state = State.check_condition
+        
+        # Check if robot stop and reach goal again in 2s
+        if self.pre_state == State.start_test and  \
+            self.current_state == State.check_condition:
+            if self.getSafeFlag("goal"):
+                if self.checkCondition(self.getSafeFlag("goal", header = True).stamp, \
```
Comment (ipa-kut) [inline araig_test_runners/scripts/test_4_static_obstacle:95]: I dont completely understand what is being checked here...

File: araig_test_runners/scripts/test_4_static_obstacle [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        if self.current_state == State.start_test and self.getSafeFlag("interrupt_test"):
+            rospy.logwarn(rospy.get_name() + ": Interrupted!!")
+            self._publishers["test_failed"].publish(self.buildNewBoolStamped(True))
+            self._publishers["test_completed"].publish(self.buildNewBoolStamped(True))
+            # !!! ui should publish interrupt_test to False after pub True, interrupt_test signal is event 
+            self._publishers["interrupt_test"].publish(self.buildNewBoolStamped(False))
+            self.current_state = State.complete_test
+
+        # If robot has stopped (because it found an obstacle), but maybe robot will still move
+        if self.current_state == State.start_test and self.getSafeFlag("goal"):
```
Comment (ipa-kut) [inline araig_test_runners/scripts/test_4_static_obstacle:87]: okay checking for oscillations. I get why you do this, but as before, the test itself should not try to compensate for oscillations. So the goal time checking is done immediately as the robot reaches goal, as that is the specification of this test. Checking for oscillations is an addendum, and if you remember our old discussion, we can just count the number of times the robot has oscillated and show that in the UI/log it.

In general, tests should not be altered to accommodate for shortcomings in the test subjects.

File: araig_test_runners/scripts/test_4_static_obstacle [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+            if self.getSafeFlag("goal"):
+                if self.checkCondition(self.getSafeFlag("goal", header = True).stamp, \
+                    self.config_param['timeout_check_goal']):
+                    rospy.loginfo(rospy.get_name() + ": Robot has reached goal and stop.")
+                    self.current_state = State.reached_goal
+            else:
+                rospy.loginfo(rospy.get_name() + ": Robot is still moving, waiting for robot reaching goal and stop")
+                self.current_state = State.start_test
+        
+        # Test complete if robot doesn't move any more and reached the goal
```
Comment (ipa-kut) [inline araig_test_runners/scripts/test_4_static_obstacle:103]: Well if the oscillations are undamped, this is basically an infinite loop :P So lets not try to compensate for it.

File: araig_test_runners/scripts/test_4_static_obstacle [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        if self.current_state == State.reached_goal:
+            rospy.loginfo(rospy.get_name() + ": Robot has reached goal, ending test in success!")
+            self._publishers["test_succeeded"].publish(self.buildNewBoolStamped(True))
+            self._publishers["test_completed"].publish(self.buildNewBoolStamped(True))
+            self.current_state = State.complete_test
+       
+        # If test complete, wait for "timeout_after_complete" s, then enable reset
```
Comment (ipa-kut) [inline araig_test_runners/scripts/test_4_static_obstacle:110]: What is the need for timeout after complete? I originally thought this was oscillation compensation, but seems to not be the case?

**Review by ipa-kut — CHANGES_REQUESTED** [review]
Seems like test 5 still needs to be refactored to the newer state-machine bound resetable format.

File: araig_test_runners/scripts/test_1_braking [hunk 7]
```diff
@@ -1,98 +1,133 @@
...
 
-            # If robot has reached max velocity
+    def main(self):
+        if self.pre_state == None and self.current_state == State.wait:
+            rospy.loginfo(rospy.get_name() + ": Waiting to start...")
+            self.pre_state = State.wait
+        
+        # Wait until /signal/ui/start_test received
+        if self.current_state == State.wait and self.getSafeFlag("start_test"):
+            self.startRecordingAndWait(3)
```
[the remark this answers is not in the document]
Comment (ipa-kut) [reply araig_test_runners/scripts/test_1_braking:64]: Wait so how do you actually trigger start of rosbagger? :thinking:

File: araig_test_runners/src/base/base_runner.py [hunk 8]
```diff
@@ -60,9 +81,31 @@ def buildNewBoolStamped(self, data = True):
         return msg
 
     def callback_for_all_bool_topics(self, msg, key):
-        self.setSafeFlag(key,msg.data)
+        self.setSafeFlag(key,msg)
 
     def startRecordingAndWait(self, duration=3):
-        self._publishers["start_recording"].publish(self.buildNewBoolStamped(True))
```
Comment (ipa-kut) [inline araig_test_runners/src/base/base_runner.py:66]: If this is removed, who tells the bagger to start recording?

File: araig_test_runners/scripts/test_1_braking (hunk 7 above)
[the remark this answers is not in the document]
Comment (ipa-kut) [reply araig_test_runners/scripts/test_1_braking:64]: TODO: Create a proper way to handle sequence of runner signals, folder creatin, loggers starting - stopping, folder renaming etc...

File: araig_test_runners/scripts/test_4_static_obstacle (hunk 4 above)
Comment (ipa-kut) [reply araig_test_runners/scripts/test_4_static_obstacle:87]: Checking for oscillations is good. So if does oscillate, then count the number, and report a third result category: test_oscillated.   
 
Verify/implement a timeout to number of oscillations that the runner can check. If more, then fail.   

Finally, log the number of oscillations. (Figure out how)

File: araig_test_runners/scripts/test_4_static_obstacle (hunk 6 above)
Comment (ipa-kut) [reply araig_test_runners/scripts/test_4_static_obstacle:110]: RWU: To protect the folder logger from being interrupted as it changes the folder name.

**Comment by ipa-kut** [comment]
This PR is superseded by https://github.com/ipa320/araig_test_stack/pull/35 and no longer required. Branch can be kept as archive.

**Pull request closed without merging**